### PR TITLE
In memory PlaceVisitStorage implementation

### LINKED ISSUE #8: Create Storage Interface
Create Storage Interface and In-memory Implementation.

### PR DESCRIPTION
**Pull request by cli1903**
note: aside from the actual implementation of the inmemory storage, I added an abstract toBuilder() method to TripModel and PlaceVisitModel to allow for transformation of objects
I also changed changePlaceVisitStatus to return false when if PlaceVisitModel object is not found in storage instead of throwing an error since we might want to add a status to a PlaceVisit that doesn't already exist for that trip

### CODE REVIEW

File: frontend/src/main/java/com/google/tripmeout/frontend/PlaceVisitModel.java [hunk 1]
```diff
@@ -8,6 +8,7 @@ public static Builder builder() {
     return new AutoValue_PlaceVisitModel.Builder();
   }
 
+  public abstract Builder toBuilder();
```
Comment (KoopaKing) [inline frontend/src/main/java/com/google/tripmeout/frontend/PlaceVisitModel.java]: Nice. I didn't realize AutoValue had that.

File: frontend/src/test/java/com/google/sps/InMemoryPlaceVisitStorageTest.java [hunk 2]
```diff
@@ -0,0 +1,333 @@
...
+    .setUserMark("if-time")
+    .setLatitude(0)
+    .setLongitude(0)
+    .build();
+
+  /** 
+   * test that a PlaceVisitAlreadyExistsException is thrown when a place that 
+   * has already been added to the storage is added again 
+   */
+  @Test(expected = PlaceVisitAlreadyExistsException.class)
```
Comment (KoopaKing) [inline frontend/src/test/java/com/google/sps/InMemoryPlaceVisitStorageTest.java]: Using `@Test(expected = ...)` is usually the wrong thing. For example if suddenly the first `addPlaceVisit` or the constructor started throwing this, this test might pass. While other tests would hopefully start failing, it might make it more difficult to root cause the error.

Instead prefer to use `org.junit.Assert.assertThrows` as so:
```java
assertThrows(PlaceVisitAlreadyExistsException.class, () -> storage.addPlaceVisit(LONDON));
```

This will also return the exception to you if you want to assert other things about it like the cause, or that a specific string was included in the message for helpfulness.

File: frontend/src/test/java/com/google/sps/InMemoryPlaceVisitStorageTest.java [hunk 3]
```diff
@@ -0,0 +1,333 @@
...
+  public void testAddSamePlaceVisit_throwsException() throws PlaceVisitAlreadyExistsException {
```
Comment (KoopaKing) [inline frontend/src/test/java/com/google/sps/InMemoryPlaceVisitStorageTest.java]: nit: Use the test naming style `methodUnderTest_situation_expectation`, so for this it would be something like `addPlaceVisit_placeAlreadyAdded_throwsException`

File: frontend/src/test/java/com/google/sps/InMemoryPlaceVisitStorageTest.java [hunk 4]
```diff
@@ -0,0 +1,333 @@
...
+
+  /**
+   * test that a PlaceNotFoundException is thrown when trying to get a place
+   * from storage when there is no place in storage with the given tripId and
+   * placeId
+   */
+  @Test(expected = PlaceVisitNotFoundException.class)
+  public void testGetPlace_notInStorage() throws PlaceVisitAlreadyExistsException, PlaceVisitNotFoundException {
+    InMemoryPlaceVisitStorage storage = new InMemoryPlaceVisitStorage();
+    storage.addPlaceVisit(LONDON);
```
Comment (KoopaKing) [inline frontend/src/test/java/com/google/sps/InMemoryPlaceVisitStorageTest.java]: nit: not sure why all these adds are necessary. I can see adding one to test the difference between the empty and the "something stored but not with this ID" cases, but this seems like overkill.

File: frontend/src/test/java/com/google/sps/InMemoryPlaceVisitStorageTest.java [hunk 5]
```diff
@@ -0,0 +1,333 @@
...
+import org.junit.runners.JUnit4;
+import java.util.List;
+import com.google.tripmeout.frontend.storage.InMemoryPlaceVisitStorage;
+import com.google.tripmeout.frontend.PlaceVisitModel;
+import com.google.tripmeout.frontend.error.PlaceVisitAlreadyExistsException;
+import com.google.tripmeout.frontend.error.PlaceVisitNotFoundException;
+
+@RunWith(JUnit4.class)
+public final class InMemoryPlaceVisitStorageTest {
+  private static final InMemoryPlaceVisitStorage NO_PLACE_VISITS = new InMemoryPlaceVisitStorage();
```
Comment (KoopaKing) [inline frontend/src/test/java/com/google/sps/InMemoryPlaceVisitStorageTest.java]: not super sure what the point of this is over just calling `new InMemoryPlaceVisitStorage()` inline is.

File: frontend/src/main/java/com/google/tripmeout/frontend/storage/InMemoryPlaceVisitStorage.java [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+import java.util.List;
+import java.util.Collection;
+import java.util.ArrayList;
+
+public class InMemoryPlaceVisitStorage implements PlaceVisitStorage {
+  // <placeId, tripId, PlaceVisitModel> 
+  Table<String, String, PlaceVisitModel> storage = Tables.synchronizedTable(HashBasedTable.create());
+ 
+  @Override
+  public void addPlaceVisit(PlaceVisitModel placeVisit) throws PlaceVisitAlreadyExistsException {
```
Comment (KoopaKing) [inline frontend/src/main/java/com/google/tripmeout/frontend/storage/InMemoryPlaceVisitStorage.java]: This isn't thread safe. If two threads call this method at the same time they can override each-others data and have neither one will get a `PlaceVisitAlreadyExistsException`.

Maybe it was a mistake to tell you to use `Tables.synchronizedTable` because it doesn't look like it has a good conditional update method, meaning you will need to do some lock management yourself.

File: frontend/src/main/java/com/google/tripmeout/frontend/storage/InMemoryPlaceVisitStorage.java [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+    PlaceVisitModel place = storage.get(placeId, tripId);
+    if (place != null) {
+      return place;
+    }
+    throw new PlaceVisitNotFoundException("PlaceVisit with id" + placeId + 
+      " not found for trip " + tripId);
+  }
+
+  @Override
+  public boolean changePlaceVisitStatus(String tripId, String placeId, String newStatus) {
```
Comment (KoopaKing) [inline frontend/src/main/java/com/google/tripmeout/frontend/storage/InMemoryPlaceVisitStorage.java]: Again, not thread safe.

**Review by KoopaKing — COMMENTED** [review]
Looks pretty good, but needs a few things addressed.

File: frontend/src/main/java/com/google/tripmeout/frontend/storage/InMemoryPlaceVisitStorage.java (hunk 6 above)
Comment (cli1903) [reply frontend/src/main/java/com/google/tripmeout/frontend/storage/InMemoryPlaceVisitStorage.java]: would it be fine to switch to a ConcurrentHashMap then? I saw that it is thread-safe but not necessarily synchronized

Comment (KoopaKing) [reply frontend/src/main/java/com/google/tripmeout/frontend/storage/InMemoryPlaceVisitStorage.java]: yeah that would be fine. The real difference is that you want something with a [compute](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#compute-K-java.util.function.BiFunction-)-like functionality which it doesn't seem like Tables have (even though it seems pretty suited to the rest of this work).

Comment (bsidhom) [reply frontend/src/main/java/com/google/tripmeout/frontend/storage/InMemoryPlaceVisitStorage.java]: I might be missing something here (or the code has changed since the initial comment as shown): why isn't `synchronized (storage)` sufficient? That work as long as _all_ accesses to the table are done in the same fashion (some changes would need to be made below). Either way, I would drop the synchronized table and use a regular table (or perhaps a map or two).

File: frontend/src/main/java/com/google/tripmeout/frontend/storage/InMemoryPlaceVisitStorage.java [hunk 8]
```diff
@@ -11,65 +11,75 @@
...
   }
 
   @Override
   public List<PlaceVisitModel> getTripPlaceVisits(String tripId) {
-    Map<String, PlaceVisitModel> placeIdPlaceMap = storage.column(tripId);
-    Collection<PlaceVisitModel> placeVisits = placeIdPlaceMap.values();
     List<PlaceVisitModel> tripPlaceVisits = new ArrayList<>();
-    for (PlaceVisitModel place: placeVisits) {
-      if (place.userMark().equals("must-see") || place.userMark().equals("if-time")) {
+    storage.forEach((placeId, tripIdMap) -> {
```
Comment (bsidhom) [inline frontend/src/main/java/com/google/tripmeout/frontend/storage/InMemoryPlaceVisitStorage.java]: Note that this suffers from the same (possible) issue as the trip storage model, unless you specifically want to use optimistic concurrency here (please see my comment in #31).

Either way, this code block is a good candidate for a stream-filter-collect flow, which might make the intent more obvious.

File: frontend/src/main/java/com/google/tripmeout/frontend/TripModel.java [hunk 9]
```diff
@@ -8,12 +8,13 @@ public static Builder builder() {
     return new AutoValue_TripModel.Builder();
   }
 
+  public abstract Builder toBuilder();
```
Comment (bsidhom) [inline frontend/src/main/java/com/google/tripmeout/frontend/TripModel.java]: Nit: please move this below the semantic fields and add a space.

File: frontend/src/main/java/com/google/tripmeout/frontend/storage/TripStorage.java [hunk 10]
```diff
@@ -12,7 +12,7 @@
    *
    * @param trip the TripModel object to add to storage
    *
-   * @throws a TripAlreadyExists exception if there is a TripModel object 
+   * @throws a TripAlreadyExists exception if there is a TripModel object
```
Comment (bsidhom) [inline frontend/src/main/java/com/google/tripmeout/frontend/storage/TripStorage.java]: FYI, having run the formatter both here and in #31 means you'll have to deal with a merge conflict on submission. Should be easy to deal with, but I wanted to point that out.

File: frontend/src/main/java/com/google/tripmeout/frontend/storage/InMemoryPlaceVisitStorage.java [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+
+    if (placesMap.remove(placeId) == null) {
+      throw new PlaceVisitNotFoundException(
+          "PlaceVisit with id" + placeId + " not found for trip " + tripId);
+    }
+  }
+
+  @Override
+  public Optional<PlaceVisitModel> getPlaceVisit(String tripId, String placeId) {
+    Map<String, PlaceVisitModel> placesMap = storage.get(tripId);
```
Comment (KoopaKing) [inline frontend/src/main/java/com/google/tripmeout/frontend/storage/InMemoryPlaceVisitStorage.java]: nit: prefer this
```java
return Optional.ofNullable(storage.get(tripId)).map(placesMap -> placesMap.get(placeId));
```

File: frontend/src/main/java/com/google/tripmeout/frontend/storage/PlaceVisitStorage.java [hunk 12]
```diff
@@ -24,40 +28,46 @@
...
+   * exist in storage
    *
    * @param tripId the id to match PlaceVisitModel object's tripId field on
    * @param placeId the id to match PlaceVisitModel object's placeId field on
-   *
-   * @throws a PlaceVisitNotFound exception if there is no PlaceVisitModel 
-   *     object in storage with the given tripId and placeId 
    */
-  void changePlaceVisitStatus(String tripId, String placeId, String newStatus);
+  boolean changePlaceVisitStatus(String tripId, String placeId, String newStatus);
```
Comment (bsidhom) [inline frontend/src/main/java/com/google/tripmeout/frontend/storage/PlaceVisitStorage.java]: Not blocking this, but why change this to a boolean from exception-throwing? Do we expect calls to `changePlaceVisitStatus` on trip/place combos that don't exist to be common? Why would that happen? Note: it's easy to forget to check a boolean return value like this (unless you enable additional static analysis tools, which I don't believe we have right now), while it's impossible to ignore a thrown exception.

The correct choice depends on the use case(s) that you have in mind.

File: frontend/src/main/java/com/google/tripmeout/frontend/storage/InMemoryPlaceVisitStorage.java [hunk 13]
```diff
@@ -0,0 +1,102 @@
...
+  // <tripId, placeId, PlaceVisitModel>
+  Map<String, Map<String, PlaceVisitModel>> storage = new ConcurrentHashMap<>();
+
+  @Override
+  public void addPlaceVisit(PlaceVisitModel placeVisit) throws PlaceVisitAlreadyExistsException {
+    Map<String, PlaceVisitModel> newPlaceMap = new ConcurrentHashMap<>();
+    newPlaceMap.put(placeVisit.placeId(), placeVisit);
+
+    Map<String, PlaceVisitModel> alreadyThereMap =
+        storage.putIfAbsent(placeVisit.tripId(), newPlaceMap);
```
Comment (bsidhom) [inline frontend/src/main/java/com/google/tripmeout/frontend/storage/InMemoryPlaceVisitStorage.java]: As a minor optimization, use `computeIfAbsent` such that we do not allocate and populate a new map if the given trip id has already been inserted. This also simplifies your overall code structure because `placeMap` _always_ contains the correct, current submap:

```java
Map<String, PlaceVisitModel> placeMap = storage.computeIfAbsent(placeVisit.tripId(), key -> new ConcurrentHashMap<>());
// placeMap is never null here and always points to the correct submap
PlaceVisitModel existingPlaceVisit = placeMap.putIfAbsent(placeVisit.placeId(), placeVisit);
if (existingPlaceVisit != null) {
  throw new ...;
}
```

Note: because the double-level insertion is not atomic (without use of external synchronization), there may be data races when deleting trips/visits. Probably not a big deal here--it depends how consistent your API needs to be. For example, this method (`addPlaceVisit`) may return successfully after a successful call to `removeTripPlaceVisits`, in which case the underlying storage no longer holds the submap we just updated here. You'll need to decide if such scenarios are acceptable.

File: frontend/src/test/java/com/google/sps/InMemoryPlaceVisitStorageTest.java [hunk 14]
```diff
@@ -0,0 +1,348 @@
...
+  }
+
+  /**
+   * test that a PlaceNotFoundException is thrown when removing a place from
+   * an empty storage
+   */
+  @Test
+  public void removePlaceVisit_emptyStorage_throwsException() throws PlaceVisitNotFoundException {
+    InMemoryPlaceVisitStorage storage = new InMemoryPlaceVisitStorage();
+    Assert.assertThrows(PlaceVisitNotFoundException.class,
```
Comment (bsidhom) [inline frontend/src/test/java/com/google/sps/InMemoryPlaceVisitStorageTest.java:157]: FYI, you can also use a static import of `assertThrows`. Up to you if that's worth it.

File: frontend/src/test/java/com/google/sps/InMemoryPlaceVisitStorageTest.java [hunk 15]
```diff
@@ -0,0 +1,348 @@
...
+    InMemoryPlaceVisitStorage storage = new InMemoryPlaceVisitStorage();
+    storage.addPlaceVisit(LONDON);
+    storage.addPlaceVisit(PARIS);
+    storage.addPlaceVisit(ROME);
+    storage.addPlaceVisit(TOKYO);
+    storage.addPlaceVisit(BEIJING);
+    storage.addPlaceVisit(TOKYO_B);
+    storage.addPlaceVisit(SEOUL);
+    storage.removeTripPlaceVisits("a");
+    List<PlaceVisitModel> tripAPlaces = storage.getTripPlaceVisits("a");
```
Comment (bsidhom) [inline frontend/src/test/java/com/google/sps/InMemoryPlaceVisitStorageTest.java:350]: In order to make tests more precise (i.e., test only the functionality you're trying to test in each case) and easier to refactor, it's best practice to only insert the place visit that you're deleting and querying for. In particular, you do not need to insert visits from trip `b` to test this. The only reason to do so is if you think that the two should interact with each other (or if you're concerned that there's a risk that place visits from unrelated trips might be returned in this case).

(This general principle applies to all test cases.)

**Review by bsidhom — COMMENTED** [review]
A few small changes, but looks good overall.

File: frontend/src/main/java/com/google/tripmeout/frontend/storage/PlaceVisitStorage.java (hunk 12 above)
Comment (cli1903) [reply frontend/src/main/java/com/google/tripmeout/frontend/storage/PlaceVisitStorage.java]: I was imagining something where we can call this function regardless of whether the user is changing the status from a previously selected status or selecting a status for the first time and that PlaceVisit objects wouldn't be stored until after the user has marked the place with a status

Comment (bsidhom) [reply frontend/src/main/java/com/google/tripmeout/frontend/storage/PlaceVisitStorage.java]: Ah, so the idea is to return true if this tis a new place visit and actually insert a new trip otherwise? I think I had missed those semantics in the in-memory implementation. Please update the comment to make this behavior as clear as possible.

It's also worth changing the method name to reflect this new behavior. For example, `addPlaceVisitOrUpdateStatus`. Or perhaps something else if you can come up with something that clearly describes the use case. Either way, this should spell out the fact that we are possibly _inserting_ a new place or updating the status of an old one.

File: frontend/src/main/java/com/google/tripmeout/frontend/storage/InMemoryPlaceVisitStorage.java [hunk 16]
```diff
@@ -0,0 +1,103 @@
...
+  @Override
+  public List<PlaceVisitModel> getTripPlaceVisits(String tripId) {
+    Map<String, PlaceVisitModel> placesMap = storage.get(tripId);
+    List<PlaceVisitModel> tripPlaceVisits = new ArrayList<>();
+    if (placesMap == null) {
+      return tripPlaceVisits;
+    }
+    for (PlaceVisitModel place : placesMap.values()) {
+      if (place != null
+          && (place.userMark().equals("must-see") || place.userMark().equals("if-time"))) {
```
Comment (bsidhom) [inline frontend/src/main/java/com/google/tripmeout/frontend/storage/InMemoryPlaceVisitStorage.java]: You should probably document these predicates in the interface method description since the behavior is not otherwise obvious to callers.

File: frontend/src/main/java/com/google/tripmeout/frontend/storage/InMemoryPlaceVisitStorage.java [hunk 17]
```diff
@@ -0,0 +1,103 @@
...
+  @Override
+  public Optional<PlaceVisitModel> getPlaceVisit(String tripId, String placeId) {
+    return Optional.ofNullable(storage.get(tripId)).map(placesMap -> placesMap.get(placeId));
+  }
+
+  @Override
+  public boolean changePlaceVisitStatus(String tripId, String placeId, String newStatus) {
+    Map<String, PlaceVisitModel> placesMap = storage.get(tripId);
+    if (placesMap == null) {
+      return false;
```
Comment (bsidhom) [inline frontend/src/main/java/com/google/tripmeout/frontend/storage/InMemoryPlaceVisitStorage.java]: Now that I understand the intent of this method:  don't you want to compute a new map value if this isn't already present? (Note: you'll still want to return false somehow, which might be tricky with `computeIfAbsent`, but there are ways you can hack around that)

If that's not the intent here, then I'm still unclear on what we're doing and you'll definitely want to expand the javadoc comment to explain what's going on.

File: frontend/src/main/java/com/google/tripmeout/frontend/storage/InMemoryPlaceVisitStorage.java [hunk 18]
```diff
@@ -0,0 +1,103 @@
...
+    }
+
+    PlaceVisitModel originalPlace = placesMap.computeIfPresent(
```
Comment (bsidhom) [inline frontend/src/main/java/com/google/tripmeout/frontend/storage/InMemoryPlaceVisitStorage.java]: As above: don't you want to insert a new value if this isn't already present?

**Review by bsidhom — DISMISSED** [review]
I'm unclear on what's supposed to happen in `changePlaceVisitStatus`; left a few comments.

File: frontend/src/main/java/com/google/tripmeout/frontend/storage/InMemoryPlaceVisitStorage.java (hunk 18 above)
Comment (cli1903) [reply frontend/src/main/java/com/google/tripmeout/frontend/storage/InMemoryPlaceVisitStorage.java]: I don't think I clearly thought this method through... Adding a new PlaceVisitModel object would make sense given the intent, but now that I've changed get to return an optional place visit object, I'm thinking now it might be better to just use get for checking whether the place is already in storage. do you think this might be better?

Comment (bsidhom) [reply frontend/src/main/java/com/google/tripmeout/frontend/storage/InMemoryPlaceVisitStorage.java]: Either approach is fine; we just want a consistent, reliable interface.

If you _do_ think you'll want to insert a new object at the same time as setting its status, then you'll need to actually update this method to do so.

If it's acceptable to require that visits be explicitly inserted before their statuses can be set, then this method should throw an exception in that case (rather than return a boolean).

Think about how callers will consume the interface. When they want to update a visit status, do you expect them to have to call `get` first to figure out if the object exists (and then insert the visit and finally update its status) or will they _know_ that the trip visit already exists? In the latter case, there's no reason to allow mutation in-line because callers are never interacting with non-existent visits. If callers need to query and conditionally insert updates in the common case, you'll probably want to support that here.

File: frontend/src/main/java/com/google/tripmeout/frontend/storage/PlaceVisitStorage.java [hunk 19]
```diff
@@ -27,31 +31,29 @@
...
-   * @param tripId the id to match PlaceVisitModel object's tripId field on
-   * @param placeId the id to match PlaceVisitModel object's placeId field on
-   *
-   * @throws a PlaceVisitNotFound exception if there is no PlaceVisitModel
-   *     object in storage with the given tripId and placeId
+   * @param placeVisit the PlaceVisitModel object to update or add
+   * @param newStatus the new status to update placeVisit's userMark field to
    */
-  void changePlaceVisitStatus(String tripId, String placeId, String newStatus);
+  boolean updateUserMarkOrAddPlaceVisit(PlaceVisitModel placeVisit, String newStatus);
```
Comment (KoopaKing) [inline frontend/src/main/java/com/google/tripmeout/frontend/storage/PlaceVisitStorage.java]: nit: If this takes a `PlaceVisitModel` why not just update the store with that value?

File: frontend/src/main/java/com/google/tripmeout/frontend/storage/InMemoryPlaceVisitStorage.java [hunk 20]
```diff
@@ -0,0 +1,107 @@
...
+import java.util.Optional;
+import java.util.concurrent.ConcurrentHashMap;
+
+public class InMemoryPlaceVisitStorage implements PlaceVisitStorage {
+  // <tripId, placeId, PlaceVisitModel>
+  Map<String, Map<String, PlaceVisitModel>> storage = new ConcurrentHashMap<>();
+
+  @Override
+  public void addPlaceVisit(PlaceVisitModel placeVisit)
+      throws PlaceVisitAlreadyExistsException, Throwable {
```
Comment (KoopaKing) [inline frontend/src/main/java/com/google/tripmeout/frontend/storage/InMemoryPlaceVisitStorage.java]: don't throw `Throwable` especially in your interface, instead check if the cause is an instance of PlaceVisitAlreadyExistsException and cast it.

```java
  } catch (RuntimeException e) {
    if (e.getCause() instanceof PlaceVisitAlreadyExistsException) {
      throw (PlaceVisitAlreadyExistsException) e.getCause();
    }
    // unexpected cause, just re-throw the unchecked runtime exception.
    throw e;
  }
```

File: frontend/src/main/java/com/google/tripmeout/frontend/storage/PlaceVisitStorage.java (hunk 19 above)
Comment (cli1903) [reply frontend/src/main/java/com/google/tripmeout/frontend/storage/PlaceVisitStorage.java]: I'm not sure this will actually be an issue in our use but just in case other fields of placeVisit differ but the tripId and placeId are the same as a PlaceVIsitModel object in storage - I want to only update the userMark field and not the other fields

Comment (cli1903) [reply frontend/src/main/java/com/google/tripmeout/frontend/storage/PlaceVisitStorage.java]: I'm not sure this will actually be an issue in our use but just in case other fields of placeVisit differ but the tripId and placeId are the same as a PlaceVIsitModel object in storage - I want to only update the userMark field and not the other fields

**Review by KoopaKing — DISMISSED** [review]

**Review by chuckluv — DISMISSED** [review]

**Review by anfee1 — DISMISSED** [review]
LGTM

**Review by anfee1 — DISMISSED** [review]

**Review by chuckluv — DISMISSED** [review]

**Review by KoopaKing — APPROVED** [review]

**Review by chuckluv — APPROVED** [review]
